rfcs/0154-Migrate-taskcluster-to-postgres.md

```diff
@@ -0,0 +1,122 @@
+* Obsoletes: None
+* Comments: [#154](https://github.com/taskcluster/taskcluster-rfcs/pull/154)
+* Initially Proposed by: @jonasfj
+
+# Summary
+
+Taskcluster currently uses Azure's
+[table storage](https://azure.microsoft.com/en-us/services/storage/tables/),
+[queue storage](https://azure.microsoft.com/en-us/services/storage/queues/),
+and [blob storage](https://azure.microsoft.com/en-us/services/storage/blobs/)
+as the data-storage backend. For example, the Queue service currently uses a
+bunch of Azure tables to handle tasks, dependencies, artifacts, and so on. This
+project involves refactoring Taskcluster to use a Postgres backend as the
+data-storage backend, instead.
+
+# Motivation
+* Azure doesn't allow joins. Postgres allows more sophisticated querying.
+* Postgres can index by other than the primary key.
+  * A shared base of indexed data can allow services to make smarter decisions and
+    save money on provisioning and data storage.
+* Azure is unreliable with frequent failures and inscrutible rate-limiting. While we are resilient
+to these errors, they do cause inefficiency and sometimes cascading failures.
+* Storing data in postgres will enable features users desire:
+  * Queue introspection (Azure doesn't allow introspection of queues).
+    * What tasks are pending right now?
+    * How many tasks are [actually pending and not cancelled](https://bugzilla.mozilla.org/show_bug.cgi?id=1434851)?
+  * Integration of task-related and worker-related status.
+    * The approach we've taken with Azure doesn't allow tables to read from
+      other tables.
+    * What tasks has this worker performed?
+  * Ad-hoc analytics.
+    * How much compute time do we use per push?
+    * What do we pay for wpt?
+  * More sophisticated dependency resolution [in the case of task failure](https://bugzilla.mozilla.org/show_bug.cgi?id=1443503#c8)
+  * Dynamically change task priorities.
+
+# Details
+
+## Transition
+
+The transition to Postgres will be done in two steps:
+1. Transfer the data from Azure to Postgres without data loss. This will not aim
+   at normalizing the data but rather prioritize a safe and easy transition.  In
+   other words, the data will be backed by Postgres, but still behaving like
+   Azure.
+2. Go fully into normalized SQL tables.
+
+In both steps, Taskcluster will get confidence in the correctness and
+performance of the implementation by first testing it in the development and
+staging environments before running this on the production environment.
+
+Upon completion of step 1, deployers of Taskcluster will need to provide a DB
+URL (maybe several) then run a script to "load" data from Azure to Postgres
+during a downtime.
+
+There will be a version of Taskcluster that supports Postgres and not Azure,
+with the previous version supporting Azure and not Postgres. So upgrading to
+that version is something that happens in a downtime.
```

Comment:
As an implementation detail, it'd be best from an ops perspective if the version bump to make this switch carried as few other changes with it as possible, preferably none.